### ISSUE TITLE
CTargetFind::getValidTarget fails check, invalid targets being assigned...

### ISSUE THREAD

**Issue by Racsr** [DarkstarProject/darkstar#2635]
...and causing the server to crash with "terminate called after throwing an instance of 'CStateInitException.'" Still trying to get to the root cause of this. I haven't been able to keep my server up for more then about 30 minutes without this crash. The exception can be thrown while in combat or even just walking through a zone (presumably a Mob goes to cast something on them selves that fails the ValidTarget check while walking by). So far I have seen crashes related to MobEntity calling both abilities and magic.

I was just debugging the issue with a Dark Spark fight and I changed the final return in targetfind.cpp from nullptr to PTarget to prevent the crash.. via debug messages I could see the call to getValidTarget reached this point when Dark Spark went to use the ability Berserk and instead of crashing targeted the PC with the effects. Same thing happened while fighting a Droma in Garlaige that used Spectral Barrier, the PC is represented as PTarget in CTargetFind which causes the crash.

This may need more testing to figure out when magic casting comes in to play as well.

**Comment by Racsr** [comment]
So I ran through Xarcabard and aggroed some Demons along the way. This run through the magic issue popped up in CMagicState::CMagicState, the "if (!CanCastSpell(PTarget))" check failed under the following events:

[18:16:17][Debug] Error 3 in CMagicState! Trying to cast 200 (waterga_ii) on 1024 () with flags 0
[18:16:52][Debug] Error 3 in CMagicState! Trying to cast 273 (sleepga) on 1024 () with flags 0
[18:17:27][Debug] Error 3 in CMagicState! Trying to cast 165 (thunder_ii) on 1024 () with flags 0
[18:18:37][Debug] Error 3 in CMagicState! Trying to cast 200 (waterga_ii) on 1024 () with flags 0

Still not sure about the random crashes while not in combat or if all these are crashes related. They all render only the "terminate called after throwing an instance of 'CStateInitException.'" error before the dsgame server dies.

By the way here's my system info:
FreeBSD 9.3-RELEASE-p28 #0 r288272+f229c79: Sat Dec 12 11:58:01 PST 2015 amd64
gcc6 (FreeBSD Ports Collection) 6.0.0 20160110 (experimental)
CPU: Intel(R) Xeon(R) CPU E3-1225 v3 @ 3.20GHz (3192.67-MHz K8-class CPU)
real memory  = 17697865728 (16878 MB)
avail memory = 16310501376 (15554 MB)

**Comment by teschnei** [comment]
mm, probably going to need a callstack if you can run it through gdb

**Comment by Racsr** [comment]
I've attached a general trace of the crash which appears fairly useless as far as dsgame goes. However, when I try and set breaks around targetfind I end up with no trace available during the crash. This is leading me to believe the entire issue could just be based on some incompatibility with my version of LIBC...

So on that lead I tried the same code base on my linux devbox:
Linux 4.2.4-1-ARCH #1 SMP PREEMPT Fri Oct 23 07:32:38 CEST 2015 x86_64 GNU/Linux
gcc (GCC) 5.3.0
CPU: Intel(R) Atom(TM) CPU  330   @ 1.60GHz (fam: 06, model: 1c, stepping: 02)
libc.so.6 (vs libc.so.7 on my FreeBSD box)

and failed to replicate the crashes.

I can try regression on my FreeBSD box another time but for now I guess this isn't really a bug with the server, just an incompatibility with either libc 7 or FreeBSD libc vs Linux glibc for reasons that are beyond me.

tl;dr Code seems fine, issues with libc/glibc.

[2635 dsgame backtrace.txt](https://github.com/DarkstarProject/darkstar/files/93766/2635.dsgame.backtrace.txt)
